File: rust/packages/blockchain/src/state.ts
import type { Transaction, TransactionType } from '@noetis/protocol';

export interface ChainAccountState {
  balances: Map<string, number>;
  escrows: Map<string, { user: string; amount: number }>;
}

export function createAccountState(): ChainAccountState {
  return { balances: new Map(), escrows: new Map() };
}

export function getBalance(state: ChainAccountState, address: string): number {
  return state.balances.get(address) ?? 0;
}

function credit(state: ChainAccountState, to: string, amount: number): void {
  state.balances.set(to, getBalance(state, to) + amount);
}

function debit(state: ChainAccountState, from: string, amount: number): boolean {
  const bal = getBalance(state, from);
  if (bal < amount) return false;
  state.balances.set(from, bal - amount);
  return true;
}

export function applyTransaction(state: ChainAccountState, tx: Transaction): boolean {
  switch (tx.type as TransactionType) {
    case 'WALLET_CREATED':
      if (tx.to) state.balances.set(tx.to, getBalance(state, tx.to));
      return true;
    case 'FAUCET_TRANSFER':
      if (tx.to) credit(state, tx.to, tx.amount);
      return true;
    case 'CURRENCY_TRANSFER':
      if (!tx.from || !tx.to) return false;
      if (!debit(state, tx.from, tx.amount)) return false;
      credit(state, tx.to, tx.amount);
      return true;
    case 'TASK_FUNDED': {
      const taskId = tx.metadata.task_id as string;
      if (!tx.from || !taskId) return false;
      if (!debit(state, tx.from, tx.amount)) return false;
      state.escrows.set(taskId, { user: tx.from, amount: tx.amount });
      return true;
    }
    case 'TASK_PAID':
    case 'TASK_REFUNDED':
      if (tx.to) credit(state, tx.to, tx.amount);
      return true;
    case 'NODE_STAKED':
      if (tx.from) return debit(state, tx.from, tx.amount);
      return false;
    case 'NODE_PENALIZED':
      if (tx.from) {
        debit(state, tx.from, Math.min(tx.amount, getBalance(state, tx.from)));
      }
      return true;
    default:
      return true;
  }
}

export function applyBlockTransactions(state: ChainAccountState, transactions: Transaction[]): boolean {
  for (const tx of transactions) {
    if (!applyTransaction(state, tx)) return false;
  }
  return true;
}

export function deriveStateFromChain(transactions: Transaction[][]): ChainAccountState {
  const state = createAccountState();
  for (const blockTxs of transactions) {
    applyBlockTransactions(state, blockTxs);
  }
  return state;
}

File: rust/crates/noetis-blockchain/src/state.rs
use noetis_protocol::{Transaction, TransactionType};
use serde_json::Value;
use std::collections::HashMap;

#[derive(Debug, Clone, Default)]
pub struct ChainAccountState {
    pub balances: HashMap<String, f64>,
    pub escrows: HashMap<String, EscrowEntry>,
}

#[derive(Debug, Clone)]
pub struct EscrowEntry {
    pub user: String,
    pub amount: f64,
}

pub fn create_account_state() -> ChainAccountState {
    ChainAccountState::default()
}

pub fn get_balance(state: &ChainAccountState, address: &str) -> f64 {
    *state.balances.get(address).unwrap_or(&0.0)
}

fn credit(state: &mut ChainAccountState, to: &str, amount: f64) {
    state
        .balances
        .insert(to.to_string(), get_balance(state, to) + amount);
}

fn debit(state: &mut ChainAccountState, from: &str, amount: f64) -> bool {
    let bal = get_balance(state, from);
    if bal < amount {
        return false;
    }
    state.balances.insert(from.to_string(), bal - amount);
    true
}

fn metadata_task_id(metadata: &HashMap<String, Value>) -> Option<String> {
    metadata
        .get("task_id")
        .or_else(|| metadata.get("taskId"))
        .and_then(|v| v.as_str())
        .map(str::to_string)
}

pub fn apply_transaction(state: &mut ChainAccountState, tx: &Transaction) -> bool {
    match tx.tx_type {
        TransactionType::WalletCreated => {
            if let Some(ref to) = tx.to {
                state.balances.entry(to.clone()).or_insert(0.0);
            }
            true
        }
        TransactionType::FaucetTransfer => {
            if let Some(ref to) = tx.to {
                credit(state, to, tx.amount);
            }
            true
        }
        TransactionType::CurrencyTransfer => {
            let (Some(ref from), Some(ref to)) = (&tx.from, &tx.to) else {
                return false;
            };
            if !debit(state, from, tx.amount) {
                return false;
            }
            credit(state, to, tx.amount);
            true
        }
        TransactionType::TaskFunded => {
            let task_id = metadata_task_id(&tx.metadata);
            let (Some(ref from), Some(task_id)) = (&tx.from, task_id) else {
                return false;
            };
            if !debit(state, from, tx.amount) {
                return false;
            }
            state.escrows.insert(
                task_id,
                EscrowEntry {
                    user: from.clone(),
                    amount: tx.amount,
                },
            );
            true
        }
        TransactionType::TaskPaid | TransactionType::TaskRefunded => {
            if let Some(ref to) = tx.to {
                credit(state, to, tx.amount);
            }
            true
        }
        TransactionType::NodeStaked => {
            if let Some(ref from) = tx.from {
                return debit(state, from, tx.amount);
            }
            false
        }
        TransactionType::NodePenalized => {
            if let Some(ref from) = tx.from {
                let amount = tx.amount.min(get_balance(state, from));
                debit(state, from, amount);
            }
            true
        }
        _ => true,
    }
}

pub fn apply_block_transactions(state: &mut ChainAccountState, transactions: &[Transaction]) -> bool {
    for tx in transactions {
        if !apply_transaction(state, tx) {
            return false;
        }
    }
    true
}

pub fn derive_state_from_chain(block_transactions: &[Vec<Transaction>]) -> ChainAccountState {
    let mut state = create_account_state();
    for block_txs in block_transactions {
        apply_block_transactions(&mut state, block_txs);
    }
    state
}
